sql/2024/02/median-image-lcp.sql

```diff
@@ -0,0 +1,82 @@
+# HTTP Archive query to get median image LCP and size in MB per image format.
+#
+# WPP Research, Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# See **TODO**
+
+SELECT
+  format,
+  APPROX_QUANTILES(lcp, 1000)[
+    OFFSET
+      (500)] AS median_lcp,
+  APPROX_QUANTILES(bytesImg, 1000)[
+    OFFSET
+      (500)] / 1024 / 1024 AS median_img_mbytes
+FROM (
+       SELECT
+         DISTINCT url,
+                  info AS version
+       FROM
+         `httparchive.technologies.2024_01_01_mobile`
+       WHERE
+           app = 'WordPress')
+       JOIN (
+  SELECT
+    url,
+    format,
+    bytesImg
+  FROM (
+         SELECT
+           pageid,
+           format
+         FROM
+           `httparchive.summary_requests.2024_01_01_mobile`
+         WHERE
+             format IN ("webp",
+                        "png",
+                        "gif",
+                        "jpg",
+                        "avif")
+         GROUP BY
+           pageid,
+           format)
+         JOIN (
+    SELECT
+      pageid,
+      url,
+      bytesImg
+    FROM
+      `httparchive.summary_pages.2024_01_01_mobile`)
+              USING
+                (pageid))
+            USING
+              (url)
+       JOIN (
+  SELECT
+    url,
+    CAST(JSON_EXTRACT_SCALAR(payload, "$['_chromeUserTiming.LargestContentfulPaint']") AS INT64) / 1000 AS lcp
```

Comment:
I'm not sure about the approach for this query. What is it that we want here? The median LCP per image format? The tricky thing with that is that LCP is for an entire page/URL, while obviously there may be multiple images per page. So how do we handle URLs that use images of multiple MIME types? Maybe the answer is simple, but it's important to clarify that.

Maybe we could just say that we group URLs by MIME type, and if a URL has images of multiple MIME types, it'll simply appear in multiple groups, and its LCP would affect all of those groups.

Potentially that is what you're trying to do here, and I'm not saying there's a problem with this query, but it's difficult to review due to all the nesting and JOINs using the legacy tables etc :) I think it'll be easier to reason about it once the query has been simplified.